geo/src/algorithm/mean_center.rs

```diff
@@ -0,0 +1,694 @@
+
+pub trait MeanCenter {
+    type Output;
+    /// Return the unweighted mean center of a geometry.
+    fn mean_center(&self) -> Self::Output;
+
+    /// Return the weighted mean center of a geometry.
+    /// The weights are cycled if there are fewer weights than geometries.
+    fn weighted_mean_center(&self, weights: &[f64]) -> Self::Output;
+}
+
+macro_rules! impl_mean_center_point {
+    ($type:ty) => {
+        impl MeanCenter for $type {
+            type Output = Point;
+            fn mean_center(&self) -> Self::Output {
+                self.centroid()
+            }
+
+            fn weighted_mean_center(&self, weights: &[f64]) -> Self::Output {
+                let mut x_sum = 0.0;
+                let mut y_sum = 0.0;
+                let mut weight_sum = 0.0;
+
+                for (coord, weight) in self.coords_iter().zip(weights.iter().cycle()) {
+                    let (xi, yi) = coord.x_y();
+                    x_sum += xi * weight;
+                    y_sum += yi * weight;
+                    weight_sum += weight;
+                }
+
+                let x = x_sum / weight_sum;
+                let y = y_sum / weight_sum;
+
+                Point::new(x, y)
+            }
+        }
+    };
+}
+
+impl_mean_center_point!(Line);
+impl_mean_center_point!(Point);
+impl_mean_center_point!(Rect);
+impl_mean_center_point!(Triangle);
+
+macro_rules! impl_mean_center_option {
+    ($type:ty) => {
+        impl MeanCenter for $type {
+            type Output = Option<Point>;
+            fn mean_center(&self) -> Self::Output {
+                self.centroid()
+            }
+
+            fn weighted_mean_center(&self, weights: &[f64]) -> Self::Output {
+                let mut x_sum = 0.0;
+                let mut y_sum = 0.0;
+                let mut weight_sum = 0.0;
+                let n = self.coords_count();
+                if n == 0 {
+                    return None;
+                }
+
+                for (coord, weight) in self.coords_iter().zip(weights.iter().cycle()) {
+                    let (xi, yi) = coord.x_y();
+                    x_sum += xi * weight;
+                    y_sum += yi * weight;
+                    weight_sum += weight;
+                }
+
+                let x = x_sum / weight_sum;
+                let y = y_sum / weight_sum;
+
+                Some(Point::new(x, y))
+            }
+        }
+    };
+}
+
+impl_mean_center_option!(LineString);
+impl_mean_center_option!(MultiPoint);
+
+impl MeanCenter for Polygon {
+    type Output = Option<Point>;
+    fn mean_center(&self) -> Self::Output {
+        self.centroid()
+    }
+
+    fn weighted_mean_center(&self, weights: &[f64]) -> Self::Output {
+        let mut x_sum = 0.0;
+        let mut y_sum = 0.0;
+        let mut weight_sum = 0.0;
+        let n = self.coords_count();
+
+        if n == 0 {
+            return None;
+        }
+
+        for ((i, coord), weight) in self.coords_iter().enumerate().zip(weights.iter().cycle()) {
```

Comment:
Hm. The blog post uses centroids here as well right?:

            xy_arr[i] = (ft_geom.Centroid().GetX() * weight, ft_geom.Centroid().GetY() * weight)